Add postgres experience

cpakulski opened the pull request:
Added logic to recognize _postgres_ flavor and bootstrap config based on template and command line parameters.

Command line parameters are specified by --templateArg. More than one templateArg may be specified. For Postgres a user must specify Endpoint, like --templateArg Endpoint=127.0.0.1

Additional parameter is InPort, which is Envoy's listening port. If not specified the value of InPort defaults to 5432.
To overwrite InPort, cmd line should contain --templateArg InPort=5433

lizan commented on pkg/cmd/run.go:
```diff
@@ -88,6 +97,29 @@ getenvoy run standard:1.11.1 -- --help
+						return fmt.Errorf("--templateArg and %s cannot be specified at the same time", envoyParam)
+					}
+				}
+				config, err := flavors.CreateConfig(key.Flavor, templateParams)
```
I did a typo running getenvoy and triggered a panic here (used `postgres/nightly` instead of `postgres:nightly`)

```
 ./getenvoy run postgres/nightly --templateArg Endpoint=127.0.0.1 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xe55895]

goroutine 1 [running]:
github.com/tetratelabs/getenvoy/pkg/cmd.NewRunCmd.func2(0xc000417b80, 0xc0004343c0, 0x1, 0x3, 0x0, 0x0)
        /home/lizan/github/tetratelabs/getenvoy/review/pkg/cmd/run.go:111 +0x205
github.com/spf13/cobra.(*Command).execute(0xc000417b80, 0xc000434360, 0x3, 0x3, 0xc000417b80, 0xc000434360)
        /home/lizan/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0x19b42c0, 0xc00009e778, 0xc00035ff78, 0x405abf)
        /home/lizan/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
        /home/lizan/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
        /home/lizan/github/tetratelabs/getenvoy/review/cmd/getenvoy/main.go:26 +0x2b
```

cpakulski replied:
I will look at this.

cpakulski replied:
Fixed.

lizan commented on pkg/flavors/flavors.go:
```diff
@@ -0,0 +1,80 @@
+
+// GetTemplate - function returns FlavorConfigTemplate structure associated
+// with flavor.
+func GetTemplate(flavor string) (FlavorConfigTemplate, error) {
```
I think we need more than flavor to infer a template, flavor:config-template isn't 1:1. For example, standard:1.15.0 should be runnable with this postgres template as well.

cpakulski replied:
I was thinking about that when was coding. I think that in addition to image types `postgres:nightly, istio:<ver> , standard:<ver>, wasm:<ver>` we should have a metadata - kind of softlink. So for example, postgres could point to standard:1.15.0, mysql could point to standard: 1.14.0. We could easily start adding flavors and keep number of builds fairly small.

cpakulski replied:
I also had https://github.com/tetratelabs/getenvoy/issues/57 issue in mind. Template repo which I added in this PR allows for default parameters, so not everything has to be specified in cmd line unless user wants to overwrite it.

lizan replied:
Yeah I think #57 is a good way to go.

> I was thinking about that when was coding. I think that in addition to image types `postgres:nightly, istio:<ver> , standard:<ver>, wasm:<ver>` we should have a metadata - kind of softlink. So for example, postgres could point to standard:1.15.0, mysql could point to standard: 1.14.0. We could easily start adding flavors and keep number of builds fairly small.

recent version of Envoy (1.13+) already reports its extension capability in admin endpoint (config_dump, https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/core/v3/base.proto#L186), so in addition to #57 we can filter based on it to see if it has enough capability.

yskopets commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+)
+
+// Define template parameter names
+const endpoint string = "endpoint"
+const inport string = "inport"
```
I think, `endpoint` and `inport` are very obscure argument names (that a user must provide on CLI).

E.g., we could prefix them to make their context more clear.

cpakulski replied:
The commands should be added to getnevoy template. Captured it here: https://github.com/tetratelabs/getenvoy/issues/97. The command should list parameters required for a template.

yskopets commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+			if !valid.IsInt(value) {
+				return nil, fmt.Errorf("Value for templateArg %s must be integer number", param)
+			}
+			flavor.inport = value
```
It's not intuitive that `CheckParams()` method with a return value actually mutates its internal state

cpakulski replied:
Changed it to use receiver.

yskopets commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+func init() {
+	// Set default values.
+	// Default values are not required to be present in cmd line.
+	flavor.inport = "5432"
```
I think, it would be more idiomatic to initialize `inport` as part of 

```
var flavor = Flavor{
    inport: 5432,
}
```

yskopets commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+		}
+	}
+
+	// Check if all required params have been found in the parameter list
+	var notFound string
+	for key, count := range required {
+		if count == 0 {
+			notFound += key + " "
+		}
+	}
```
This is a very generic way to validate parameters.

Especially, when we have only 2 of them :)

May be, we could avoid the `required ` map and `for` loop ?

cpakulski replied:
I will see if I can move it to `flavors` package and keep it generic. It could be useful when we expand flavors.

cpakulski replied:
Simplified it when I reworked logic to accommodate multiple endpoints.

yskopets commented on pkg/cmd/run.go:
```diff
@@ -162,3 +204,24 @@ func controlplaneFunc() func(r *envoy.Runtime) {
+	if err != nil {
+		return "", err
+	}
+	return "--config-path " + path, nil
```
Why do we concatenate arguments into a single string instead of returning a string array ?

Does `Envoy` understand correctly `'--config-path /path/to/config'` as a single string ?

lizan replied:
Yes

yskopets commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+
+// CheckParams verifies that passed template arguments are correct and
+// are sufficient for creating a valid config from template.
+func (Flavor) CheckParams(params map[string]string) (interface{}, error) {
```
Based on the contents of this method, may be it would be better to refactor it into `Parse()` function to avoid non-intuitive mutation ?

cpakulski replied:
Changed name CheckParseParams and refactored to take receiver and update receiver's state instead of global.

lizan commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+  address:
+    socket_address:
+      address: 0.0.0.0
+      port_value: 8001`
```
Ideally this should be templated and default to 15000. We have some debug features that rely on this.

lizan replied:
Yes let's open an issue.

cpakulski replied:
Create issue for this: https://github.com/tetratelabs/getenvoy/issues/96.

musaprg commented:
`getenvoy run postgres:nightly --templateArg endpoint=127.0.0.1` got errors below

```
[2020-07-22 18:43:38.479][39037][critical][main] [external/envoy/source/server/server.cc:98] error initializing configuration '/home/mssn/.getenvoy/configs/postgres.yaml': Proto constraint validation failed (BootstrapValidationError.StaticResources: ["embedded message failed validation"] | caused by StaticResourcesValidationError.Listeners[i]: ["embedded message failed validation"] | caused by ListenerValidationError.Address: ["embedded message failed validation"] | caused by AddressValidationError.SocketAddress: ["embedded message failed validation"] | caused by field: "port_specifier", reason: is required): static_resources {
  listeners {
    name: "postgres_listener"
    address {
      socket_address {
        address: "0.0.0.0"
      }
    }
  }
}

[2020-07-22 18:43:38.479][39037][info][main] [external/envoy/source/server/server.cc:645] exiting
Proto constraint validation failed (BootstrapValidationError.StaticResources: ["embedded message failed validation"] | caused by StaticResourcesValidationError.Listeners[i]: ["embedded message failed validation"] | caused by ListenerValidationError.Address: ["embedded message failed validation"] | caused by AddressValidationError.SocketAddress: ["embedded message failed validation"] | caused by field: "port_specifier", reason: is required): static_resources {
  listeners {
    name: "postgres_listener"
    address {
      socket_address {
        address: "0.0.0.0"
      }
    }
  }
}
```

it seems to be due to the broken config `.getenvoy/configs/postgres.yml`.

```
$ cat ~/.getenvoy/configs/postgres.yaml
static_resources:
  listeners:
  - name: postgres_listener
    address:
      socket_address:
        address: 0.0.0.0
        port_value: %
```

musaprg commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+// and stores config data specific to Postgres template.
+type Flavor struct {
+	// Location of the postgres server
+	endpoint string
+	// Envoy's listener port
+	inport string
```
These fields should be exported to parse them by `text/template` packages.

cpakulski replied:
I just noticed that too. Thanks. I had originally Endpoint and InPort, but changed to lowercase because of linter.

cpakulski commented on pkg/flavors/flavors.go:
```diff
@@ -0,0 +1,80 @@
+		return "", fmt.Errorf("Supplied template for flavor %s is incorrect", flavor)
+	}
+	var buf bytes.Buffer
+	tmpl.Execute(&buf, data) //nolint
```
Because the template is hardcoded in the source code and all input parameters for substitution are strings we have full control over input parameters. It would be different situation if the template was supplied by a user and could contain syntax errors.

cpakulski replied:
Added code to check the result.

cpakulski commented on pkg/cmd/run.go:
```diff
@@ -162,3 +204,24 @@ func controlplaneFunc() func(r *envoy.Runtime) {
+		return "", err
+	}
+	// Save config in getenvoy directory
+	path, err := runtime.SaveConfig(flavor, config)
```
I think we should generate permanent config files and keep overwriting them. The reason is that a user may bootstrap basic Postgres (or any other flavor) config, copy it out, modify to their purpose and run getenvoy with custom config.

yskopets replied:
My concern is that such a config file will be constantly overwritten.

So, although the file is there, a user cannot really rely on it. We might even need to warn a user to make a copy of this file if he doesn't want to lose it.

yskopets replied:
Another question, we save the file but don't tell the user about it.

What is the value of that file then ? (from a user perspective)

yskopets replied:
Another thing to consider, there must be a way for a user to learn more about configuration templates supported by `getenvoy`, their arguments, usage examples, etc.

Let's say something like `getenvoy template list | show <template_name> | render <template_name>`

Then a user could generate a config file by running `getenvoy template render <template_name> --arg endpoint=host1:5432`.

And we don't need to store configs.

cpakulski replied:
Hmm, I have to rethink it. Templates are derived from image flavor, like postgres:nightly, standard:1.15, istio:1.15. Based on the prefix, the corresponding template is found. Right now there is only postgres, but other ones like mysql can be added. Image type is also chosen based on prefix and it is guaranteed that image is going to work with a config template. If we allow specifying image type and template separately, it could lead to conflicts.

cpakulski replied:
Captured it in new issue: https://github.com/tetratelabs/getenvoy/issues/97. Closing it for now.

cpakulski commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+		switch param {
+		case endpoint:
+			required[param]++
+			flavor.endpoint = value
```
Modified it to use method's receiver to update internal state and not use global variable. Thanks for catching it.

cpakulski commented on pkg/cmd/run.go:
```diff
@@ -153,6 +164,37 @@ func validateRequiresBootstrap() error {
 	return nil
 }
 
+func validateTemplateArg() error {
```
Thanks. It drastically simplified the code. Thanks.

cpakulski commented on pkg/cmd/run.go:
```diff
@@ -153,6 +164,37 @@ func validateRequiresBootstrap() error {
+	if err := validateBootstrap(); err != nil {
+		return err
+	}
+	templateParams = make(map[string]string)
+	if err := validateTemplateArg(); err != nil {
+		return err
+	}
```
I did not touch that part. It was like that before.

cpakulski commented on pkg/binary/interface.go:
```diff
@@ -39,6 +39,7 @@ type Runner interface {
 	DebugStore() string
 	SetStdout(w io.Writer)
 	SetStderr(w io.Writer)
+	SaveConfig(name string, config string) (string, error)
```
Hmm. I need location of $HOME/.getenvoy and it is in Runtime. Let me think how to access if clearly. Suggestions?

yskopets replied:
In `getenvoy-wasm` code of `NewRuntime()` has been refactored.

`$HOME/.getenvoy` is available as a global variable. See https://github.com/tetratelabs/getenvoy-wasm/blob/master/pkg/common/home.go

lizan replied:
add a TODO here, we can address that when we refactor NewRuntime.

cpakulski commented:
@musaprg, @yskopets, @lizan I addressed most comments. The remaining issues are:

1. endpoint 
- should we stick to only one endpoint in the cluster or allow for multiple
- right now it must be IP address

2. conflict with -c/--config-path parameters which may be specified by a user. Not sure that is the best approach to this.
Basically when templateArgs are specified, getenvoy creates config file and appends it to the existing cmd line parameters. We can end up with potentially conflicting settings.

yskopets commented on pkg/cmd/run.go:
```diff
@@ -53,18 +56,13 @@ getenvoy run ./envoy -- --config-path ./bootstrap.yaml
 getenvoy run standard:1.11.1 -- --help
+
+# Run with Postgres specific configuration bootstrapped
+getenvoy run postgres:nightly --templateArg Endpoint=127.0.0.0 --templateArg InPort=1114
```
It's unexpected to see argument names starting from a capital letter: `Endpoint` and `InPort`.

I think, this is an implementation detail leaking into user interface.

I think, users should provide `endpoint` and `ingress.port`.

yskopets commented on pkg/cmd/run.go:
```diff
@@ -114,6 +125,8 @@ getenvoy run standard:1.11.1 -- --help
 		"(experimental) location of Envoy's access log server <host|ip:port> (requires bootstrap flag)")
 	cmd.Flags().StringVar(&mode, "mode", "",
 		fmt.Sprintf("(experimental) mode to run Envoy in <%v> (requires bootstrap flag)", strings.Join(envoy.SupportedModes, "|")))
+	cmd.Flags().StringToStringVarP(&templateArgs, "templateArg", "", map[string]string{},
```
nit: suffix `*P` in the name `StringToStringVarP` means that the option should have a single letter variant. However, you provide `""`.

In this case it's more appropriate to use `StringToStringVar` (without `*P` suffix)

cpakulski replied:
Fixed. Thank you.

yskopets commented on pkg/flavors/flavors.go:
```diff
@@ -0,0 +1,80 @@
+
+// FlavorConfigTemplate - interface to individual flavors.
+type FlavorConfigTemplate interface {
+	CheckParseParams(params map[string]string) error
```
Sorry if I was ambiguous.

I was proposing to rename `CheckParams()` into `Parse<something>()`.

`CheckParseParams()` is too complicated for me :)

yskopets commented on pkg/flavors/flavors.go:
```diff
@@ -75,6 +75,6 @@ func CreateConfig(flavor string, params map[string]string) (string, error) {
 	}
 	var buf bytes.Buffer
-	tmpl.Execute(&buf, data) //nolint
+	tmpl.Execute(&buf, flavorData) //nolint
```
Why asymmetry in the way errors get handled?

Error from `tmpl.Parse()` is reported.

Error from `tmpl.Execute()` is silenced.

cpakulski replied:
Moved this code to postgres flavor and added error checking for both functions.

yskopets commented:
> should we stick to only one endpoint in the cluster or allow for multiple ?

My vote is for supporting multiple endpoints. A user will have to provide a `,` separated value, e.g. `host1:5432,host2:5432`

> right now it must be IP address

If I understand correctly, `LOGICAL_DNS` fixes the problem. Let's use it instead of `STATIC`.

> conflict with -c/--config-path parameters which may be specified by a user. Not sure that is the best approach to this.
Basically when templateArgs are specified, getenvoy creates config file and appends it to the existing cmd line parameters. We can end up with potentially conflicting settings.

I think, it's reasonable behaviour. Even if it might fail occasionally.

yskopets commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,124 @@
+        - endpoint:
+            address:
+              socket_address:
+                address: {{ .endpoint}} 
```
> but then we should check that only one domain has been specified.

Why ? Is it a constraint of `Envoy` ?

cpakulski replied:
AFAIK for dns clusters, Envoy just knows one DNS name and DNS server manages multiple endpoints.

lizan replied:
Yes only one endpoint for DNS clusters, it's up to management server to covert multiple DNS names to EDS for multiple DNS entries.

cpakulski commented:
> > right now it must be IP address
> 
> If I understand correctly, LOGICAL_DNS fixes the problem. Let's use it instead of STATIC.

The difference between LOGICAL and STRICT DNS is that in STRICT mode, DNS server returns set of IP addresses and Envoy should distribute load among them. LOGICAL DNS has kind of load balancing implemented and each time Envoy ask it returns different IP address, so load balancing is implemented in DNS, not Envoy. But AFAIK it still requires domain name, not endpoint's IP.

It will complicate the logic, but I will implement it as follows:
- if endpoints are IP addresses, use STATIC and create multiple endpoints
- if endpoint is name, use STRICT_DNS and verify that only one endpoint is specified.

lizan commented:
> > > right now it must be IP address
> > 
> > 
> > If I understand correctly, LOGICAL_DNS fixes the problem. Let's use it instead of STATIC.
> 
> The difference between LOGICAL and STRICT DNS is that in STRICT mode, DNS server returns set of IP addresses and Envoy should distribute load among them. LOGICAL DNS has kind of load balancing implemented and each time Envoy ask it returns different IP address, so load balancing is implemented in DNS, not Envoy. But AFAIK it still requires domain name, not endpoint's IP.
> 
> It will complicate the logic, but I will implement it as follows:
> 
> * if endpoints are IP addresses, use STATIC and create multiple endpoints
> * if endpoint is name, use STRICT_DNS and verify that only one endpoint is specified.

For now I think we can do STRICT_DNS with only one endpoint, we can make this improvement as a follow up. As this is targeted for local devs.

cpakulski commented:
Reworked logic in flavor processing.

Multiple endpoints may be specified, but they must be the same type. 

If all endpoints are IP addresses:
`getevoy run postgres:nightly --templateArg endpoint=127.0.0.1:1234,128.0.0.1:1234`
the STATIC cluster will be created.

If all endpoints are hostnames:
`getevoy run postgres:nightly --templateArg endpoint=postgres1:1234,postgres2:1234`
strict_dns cluster will be created.

Mixing types will result in error:
`getevoy run postgres:nightly --templateArg endpoint=127.0.0.1:1234,postgres:1234`

cpakulski commented:
I think I addressed all comments. The remaining issues are:

- input params: I named them `endpoint` and `inport`. @yskopets suggested `ingress.port`. I like the naming, but the dot suggests that there is multi-level set of params and port is just one of them.

- config file is still saved in .getenvoy/configs directory. It is just overwritten each time getenvoy is run. I personally like the fact that I can copy this file out and use it as starting point for more complex config. Using temporary file would not allow for that.

cpakulski commented on pkg/cmd/run.go:
```diff
@@ -53,18 +56,13 @@ getenvoy run ./envoy -- --config-path ./bootstrap.yaml
 getenvoy run standard:1.11.1 -- --help
+
+# Run with Postgres specific configuration bootstrapped
+getenvoy run postgres:nightly --templateArg endpoint=127.0.0.1:5432 --templateArg inport=5555
```
Done.

cpakulski commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,264 @@
+)
+
+// Define template parameter names
+const endpoint string = "endpoint"
```
Renamed to "endpoints".

cpakulski commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		return "", err
+	}
+	config += endpointsConfig
```
I will open the issue for this once it ie merged. It is internal change not affecting a user. It is rather bigger change both on code and test side.

cpakulski commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,264 @@
+
+	// Check if all required params have been found in the parameter list
+	if len(f.endpoints) == 0 {
+		return fmt.Errorf("endpoint parameter must be specified")
```
I rephrased it. It started with lowercase because it refers to command line parameter "endpoint". I changed it to "Parameter endpoints must be specified:.

cpakulski commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,264 @@
+	}
+
+	// This is just IP address or domain name.
+	isIP := (net.ParseIP(host) != nil)
```
I am still thinking C++. Removed ().

cpakulski commented on pkg/flavors/postgres/postgres_test.go:
```diff
@@ -0,0 +1,213 @@
+	}
+
+	for _, testCase := range input {
+		testFlavor.endpoints = testFlavor.endpoints[:0]
```
Yes, good idea.

cpakulski commented on pkg/flavors/postgres/postgres_test.go:
```diff
@@ -0,0 +1,213 @@
+		if err != nil {
+			continue
+		}
+		if *parsed != testCase.parseResult {
```
I will do it when I change template execution to use `range`.

cpakulski commented on pkg/flavors/postgres/postgres_test.go:
```diff
@@ -0,0 +1,213 @@
+
+		// Scan created config for input params
+		for _, item := range testCase.output {
+			if !strings.Contains(endpointsConfig, item) {
```
I will do it when I change template execution to use `range`.

cpakulski commented on pkg/flavors/postgres/postgres_test.go:
```diff
@@ -0,0 +1,213 @@
+
+		// Scan created config for input params
+		for _, item := range testCase.output {
+			if !strings.Contains(endpointsConfig, item) {
```
I will do it when I change template execution to use `range`.

cpakulski commented on pkg/flavors/flavors.go:
```diff
@@ -0,0 +1,62 @@
+	"fmt"
+)
+
+// FlavorConfigTemplate - interface to individual flavors.
```
Agree. Corrected.

cpakulski commented on pkg/flavors/postgres/postgres.go:
```diff
@@ -0,0 +1,264 @@
+			}
+			f.InPort = value
+		default:
+			fmt.Printf("Ignoring unrecognized template parameter: %s", param)
```
Changed to use tetratelabs/log.

cpakulski commented on pkg/cmd/run.go:
```diff
@@ -162,3 +204,24 @@ func controlplaneFunc() func(r *envoy.Runtime) {
+// Envoy command line to invoke a proper config.
+func processTemplateArgs(flavor string, args []string, templateParams map[string]string, runtime *envoy.Runtime) (string, error) {
+	for _, envoyParam := range args {
+		if strings.HasPrefix(envoyParam, "--config") {
```
I think I will omit this check all together. Envoy will not launch and exit then two `-c` options or one `-c` and one `--config-path` are specified. What if Envoy ads new cmd params in the future and break whatever logic we build here? WDYT?

yskopets replied:
Ok. I agree that it's enough to get started

cpakulski commented:
Removed code checking if a user also specified config file. If it is the case, Envoy will complain with `-c (--config-path) -- Argument already set!` message.
